DOC Prepare for docs site (#981)
tractorcow-farm/silverstripe-fluent: docs/en/04_dataobjects.md

Reviewer: Looks like this was moved from the `docs/en/dataobjects.md` file - I will try to compare them and not make too many suggestions where you haven't changed the existing docs, but I might miss a few cases.

If I'm asking for changes to stuff that was already there and is out of scope feel free to let me know.

---

> title: Using DataObjects with Fluent

Reviewer: ```suggestion
title: Using DataObject models with Fluent
```

Author: Updated

> summary: Configuring DataObjects for use with Fluent, including CMS fields, extensions, and versioned objects.

Reviewer: ```suggestion
summary: Configuring DataObject models for use with Fluent, including CMS fields, extensions, and versioned objects.
```

Author: Updated

icon: database
---

> # Using dataObjects with fluent

Reviewer: ```suggestion
# Using `DataObject` models with fluent
```

Author: Updated

## CMS fields

With each of the possible extensions that come with Fluent, it's necessary to configure your DataObjects to
correctly delegate your `$fields` to extensions. A different approach may be needed for each DataObject, depending
on how it generates it's fields.

In all cases, it is necessary to ensure that the `updateCMSFields()` extension method is called once (and
only once) on each `FieldList` object.

### Default field scaffolder

> If using a default field scaffolder (such as [`SiteTree::getCMSFields()`](api:SilverStripe\CMS\Model\SiteTree::getCMSFields()) or `DataObject::getCMSFields`)

Reviewer: ```suggestion
If using a default field scaffolder (such as [`DataObject::getCMSFields()`](api:SilverStripe\ORM\DataObject::getCMSFields()))
```
It's the same method, so only should be mentioned once (especially if we're only linking to one)

Author: Updated

> it is necessary to know that these methods both, by default, will call the

Reviewer: ```suggestion
it is necessary to know that these methods, by default, will call the
```

Author: Updated

`extend('updateCMSFields', $fields)` extension before returning.

If you wish to add any translated form fields to the result of this call, then you should use [`DataObject::beforeUpdateCMSFields()`](api:SilverStripe\ORM\DataObject::beforeUpdateCMSFields()):

```php
namespace App\Pages;

> class MyPage extends SiteTree

Reviewer: ```suggestion

use Page;

class MyPage extends Page
```
Needs `use` statement and shouldn't directly subclass `SiteTree`

Author: Updated

{

>     public function getCMSFields()

Reviewer: ```suggestion
    // ...
    public function getCMSFields()
```
Indicates that we're skipping some best practice (e.g. `table_name`) for the sake of brevity.

Author: Updated

    {
        // Adding the Description field early will allow FluentField to decorate this with the appropriate
        // CSS classes.
        $this->beforeUpdateCMSFields(function ($fields) {

>             $fields->addFieldToTab('Root.Main', new TextField('Description'));

Reviewer: ```suggestion
            $fields->addFieldToTab('Root.Main', TextField::create('Description'));
```
Please also add the `use` statement for `TextField`

Author: Updated

        });

        // The result of this call will have had beforeUpdateCMSFields then updateCMSFields called on it
        $fields = parent::getCMSFields();
        $fields->removeByName('Content', true);
        return $fields;
    }
}
```

### Explicit field generation

If explicitly generating your [`FieldList`](api:SilverStripe\Forms\FieldList), simply make sure to call the appropriate extension hook prior to returning.

```php
namespace App\Models;

> class MyObject extends DataObject

Reviewer: Please add the `use` statements for `DataObject`, `FieldList`, `TextField`, and `TextareaField`.

Author: Updated

{
    public function getCMSFields()
    {
        // Note the absence of any parent::getCMSFields
        $fields = FieldList::create(
            TextField::create('Title', 'Title', null, 255),
            TextareaField::create('Description')
        );

        // This line is necessary, and only AFTER you have added your fields
        $this->extend('updateCMSFields', $fields);
        return $fields;
    }
}
```

### Extensions

> As mentioned in the CMS Fields heading, it is necessary to ensure that the `updateCMSFields()` extension method is called

Reviewer: ```suggestion
As mentioned in the [CMS Fields section](#cms-fields), it is necessary to ensure that the `updateCMSFields()` extension method is called
```

Author: Updated

once (and only once) on each `FieldList` object. Fields added by other extensions will not have the necessary
decorations applied to it by the [`FluentExtension`](api:TractorCow\Fluent\Extension\FluentExtension), unless it happens to be applied before `FluentExtension`. Since
extension order cannot be guaranteed, this poses a problem.

You can circumvent this issue by extending `FluentExtension` from your custom extension and then calling
`parent::updateCMSFields($fields)` at the end of your method. This way it is actually one call to `updateCMSFields()`
using inheritance rather than two different extensions:

```php
namespace App\Extensions;

use TractorCow\Fluent\Extension\FluentExtension;

class PlayerExtension extends FluentExtension
{

>     public function updateCMSFields(FieldList $fields)

Reviewer: Please add the `use` statement for `FieldList`

Author: Updated

    {
        $fields->removeByName('SomeIrrelevantField');
        parent::updateCMSFields($fields);
    }
}
```

#### Versioned extension

> **Important:** If you're applying the [`FluentVersionedExtension`](api:TractorCow\Fluent\Extension\FluentVersionedExtension) to a versioned DataObject, you will need to ensure that
> it is applied *after* the [`Versioned`](api:SilverStripe\Versioned\Versioned) extension. You can control this with an `after: '#versionedfiles'` rule in your
> YAML title block. An example configuration block might look like this:

Reviewer: ```suggestion
> [!IMPORTANT]
> If you're applying the [`FluentVersionedExtension`](api:TractorCow\Fluent\Extension\FluentVersionedExtension) to a versioned DataObject, you will need to ensure that
> it is applied *after* the [`Versioned`](api:SilverStripe\Versioned\Versioned) extension. You can control this with an `after: '#versionedfiles'` rule in your
> YAML title block. An example configuration block might look like this:
```

Author: Updated
